### ISSUE TITLE
allow 'empty' CDNSKEY that signals DoT without pinning?

### ISSUE THREAD

**Issue by Habbie** [PowerDNS/parent-signals-dot#6]
(suggested by @mind04)

Reasons:
* people may want opportunistic encryption where they promise port 853 works (or fails fast with TCP RST on -some- of their NSes) but don't want the hassle of key management
* it allows room for future extensions (TLSA on `_853._tcp.example._dnssec.com`, as per Paul Vixie in recent dns-operations threads, as a rough example, although it would not work in this form) that could do more than pin the key

This would take the form of `example.com DNSKEY 0 3 TBD -`, where `-` represents a zero length Public Key field inside the (C)DNSKEY.

I am not super convinced by either argument right now, but I do like leaving room now for those future extensions, whatever they might be.

**Comment by RobinGeuze** [comment]
This would make validating the key a bit heavier since you'd also have to check a zero length pubkey against the DS, since that will be still a normal hash. Doesn't it make more sense to add a separate algorithm for it, that would keep the recursor logic much simpler, eg. "if you see this algo, just do DoT, don't bother validating"

**Comment by mind04** [comment]
A second algorithm for the non validate case will complicate things al lot. For example what will happen when both algorithms are present in a single delegation.

It is unlikely that the empty hash will consume more resources. You only need to calculate this hash if non of the DS records match the certificate. In a zone where all certificates are pinned, this will never happen.
In a zone where only the empty DS is present you need to calculate a hash for the first certificate and one for the empty DS. Once the empty DS is found there is no need to calculate hashes for the remaining certificates, and in the end this will save resources.
Only with a mixture of pinned and unpinned certificates this will result in one additional hash calculation. Since this case is somewhat unlikely, and the price is at most calculating one hash, I think the additional algorithm for the non validate case is overkill.

**Comment by RobinGeuze** [comment]
Tbh it feels like a really ugly solution, and I find neither argument very compelling.

IMO if we put this inside the draft we should disallow mixing pinned and unpinned since there is no benefit to pinning if you also have unpinned.

**Comment by RobinGeuze** [comment]
I've been thinking about this a bit further and signalling without pinning seems a bit pointless.

The reason is simple, you want privacy, however without the pinning there is no more privacy than a normal plain text connection to someone that is MiTM. If there is no cert checking at all he can just set up his own secure connection.

The case where CA checking is used is not much better. Since it relies on hostnames its pretty much insecure, since NS record are unsigned and therefore could be changed by the attacker to allow him to server his own cert.

**Comment by Habbie** [comment]
Based on discussion between me and Robin, I think we should do -00 with only the pinning variant. We'll see if the dprive WG thinks that is too strong.

**Comment by chantra** [comment]
On the operational side, I can see as beneficial to enable DoT without possibly shooting oneself on the foot if the DS is messed up. But maybe in this case, it better be an attribute of the DS record that signals whether DoT should be opportunistic or strict? This may alleviate the special casing on the implementation side?

**Comment by Habbie** [comment]
Because many registries take (C)DNSKEY from the registrants, we can't set any attribute on the DS. Everything we do has to be gated through the question 'is the hash of this pseudo DNSKEY record in the DS set?'. Under that restriction, we can do DNSKEY flags (but that could quickly get out of control combinatorially), but I think the empty key would be the cleaner approach.

~That said, I'd still like to submit -00 with just pinning, if you agree?~ did that :)

**Comment by RobinGeuze** [comment]
Should we raise this as a talk point on IETF?

**Comment by chantra** [comment]
Yes, it is probably worth it to have a list of items that were raised on the mailing list and would need further discussion.
